File: packages/provide-uterm-go/pty/captureconnector.go
//
// SPDX-FileCopyrightText: Copyright (c) 2025-2026 provide.io llc. All rights reserved.
// SPDX-License-Identifier: AGPL-3.0-or-later
//

package pty

import (
	"context"
	"fmt"
	"net"
	"strings"
	"sync"
)

// captureBufferCap bounds the CaptureConnector rendered buffer. Port of
// capture_connector.py's 65536 cap.
const captureBufferCap = 65536

// captureConnectorKeys is the accepted connector_config key set. Port of
// capture_connector.py _VALID_CONFIG_KEYS.
var captureConnectorKeys = map[string]struct{}{
	"socket_path": {}, "cols": {}, "rows": {}, "connect_timeout_s": {},
	"input_mode": {}, "stdin_socket_path": {},
}

// CaptureConnector observes an LD_PRELOAD/DYLD-captured shell over a Unix socket
// (no process is forked). connector_type = "pty_capture". Port of
// capture_connector.CaptureConnector.
//
// Only CHANNEL_STDOUT contributes to the visible screen; CHANNEL_STDIN and
// CHANNEL_CONNECT frames are recorded for the analysis log.
type CaptureConnector struct {
	sessionID       string
	displayName     string
	socketPath      string
	cols            int
	rows            int
	connectTimeout  float64
	stdinSocketPath string

	mu          sync.Mutex
	capture     *CaptureSocket
	connected   bool
	buffer      string
	pending     string
	connectLog  []string
	stdinCount  int
	stdinWriter net.Conn
}

// NewCaptureConnector validates config and builds a capture connector. Port of
// CaptureConnector.__init__.
func NewCaptureConnector(sessionID, displayName string, config map[string]any) (*CaptureConnector, error) {
	if err := checkUnknownKeys("CaptureConnector", config, captureConnectorKeys); err != nil {
		return nil, err
	}
	if _, ok := config["socket_path"]; !ok {
		return nil, fmt.Errorf("CaptureConnector requires 'socket_path' in connector_config")
	}
	socketPath := fmt.Sprintf("%v", config["socket_path"])
	stdinPath, _ := optString(config, "stdin_socket_path")

	timeout := 5.0
	switch t := config["connect_timeout_s"].(type) {
	case float64:
		timeout = t
	case int:
		timeout = float64(t)
	}

	return &CaptureConnector{
		sessionID:       sessionID,
		displayName:     displayName,
		socketPath:      socketPath,
		cols:            coerceIntOr(config["cols"], 80),
		rows:            coerceIntOr(config["rows"], 24),
		connectTimeout:  timeout,
		stdinSocketPath: stdinPath,
	}, nil
}

// Start binds the capture socket. Port of CaptureConnector.start.
func (c *CaptureConnector) Start(ctx context.Context) error {
	cs, err := NewCaptureSocket(c.socketPath)
	if err != nil {
		return err
	}
	if err := cs.Start(); err != nil {
		return err
	}
	c.mu.Lock()
	c.capture = cs
	c.connected = true
	c.mu.Unlock()
	return nil
}

// Stop closes the stdin forwarder and the capture socket. Port of
// CaptureConnector.stop.
func (c *CaptureConnector) Stop(ctx context.Context) error {
	c.closeStdinWriter()
	c.mu.Lock()
	cs := c.capture
	c.capture = nil
	c.connected = false
	c.mu.Unlock()
	if cs != nil {
		return cs.Stop()
	}
	return nil
}

// IsConnected reports the connection state. Port of CaptureConnector.is_connected.
func (c *CaptureConnector) IsConnected() bool {
	c.mu.Lock()
	defer c.mu.Unlock()
	return c.connected
}

// PollMessages drains all immediately-available capture frames, updating the
// buffer/analysis state, and returns a single "term" frame with the newly
// captured stdout (or nil). Port of CaptureConnector.poll_messages.
func (c *CaptureConnector) PollMessages() []Frame {
	c.mu.Lock()
	defer c.mu.Unlock()
	if !c.connected || c.capture == nil {
		return nil
	}
	changed := false
	for {
		frame, ok := c.capture.ReadNowait()
		if !ok {
			break
		}
		switch frame.Channel {
		case ChannelStdout:
			// Normalize bare \n → \r\n: capture bypasses the PTY ONLCR driver,
			// so xterm.js would otherwise advance the cursor without a CR.
			text := decodeReplace(frame.Data)
			text = strings.ReplaceAll(text, "\r\n", "\n")
			text = strings.ReplaceAll(text, "\n", "\r\n")
			c.buffer += text
			if len(c.buffer) > captureBufferCap {
				c.buffer = c.buffer[len(c.buffer)-captureBufferCap:]
			}
			c.pending += text
			changed = true
		case ChannelStdin:
			c.stdinCount++
		case ChannelConnect:
			c.connectLog = append(c.connectLog, decodeReplace(frame.Data))
			if len(c.connectLog) > 100 {
				c.connectLog = c.connectLog[len(c.connectLog)-100:]
			}
		}
	}
	if changed && c.pending != "" {
		data := c.pending
		c.pending = ""
		return []Frame{{"type": "term", "data": data}}
	}
	return nil
}

// HandleInput forwards keystrokes to the stdin socket when configured. Port of
// CaptureConnector.handle_input.
func (c *CaptureConnector) HandleInput(ctx context.Context, data string) []Frame {
	if c.stdinSocketPath != "" {
		c.forwardStdin([]byte(data))
	}
	return nil
}

// forwardStdin lazily connects the stdin socket, writes, and reconnects+retries
// once on error. Port of CaptureConnector._forward_stdin.
func (c *CaptureConnector) forwardStdin(data []byte) {
	for attempt := 0; attempt < 2; attempt++ {
		c.mu.Lock()
		if c.stdinWriter == nil {
			conn, err := net.Dial("unix", c.stdinSocketPath)
			if err != nil {
				c.mu.Unlock()
				return
			}
			c.stdinWriter = conn
		}
		w := c.stdinWriter
		c.mu.Unlock()

		if _, err := w.Write(data); err == nil {
			return
		}
		c.closeStdinWriter()
	}
}

// closeStdinWriter closes the stdin forwarding connection, swallowing errors.
// Port of CaptureConnector._close_stdin_writer.
func (c *CaptureConnector) closeStdinWriter() {
	c.mu.Lock()
	w := c.stdinWriter
	c.stdinWriter = nil
	c.mu.Unlock()
	if w != nil {
		_ = w.Close()
	}
}

// HandleControl is a no-op for the capture connector. Port of
// CaptureConnector.handle_control.
func (c *CaptureConnector) HandleControl(action string) []Frame { return nil }

// GetSnapshot returns a fresh snapshot frame. Port of CaptureConnector.get_snapshot.
func (c *CaptureConnector) GetSnapshot() Frame {
	c.mu.Lock()
	defer c.mu.Unlock()
	return c.snapshotLocked()
}

// SetMode re-advertises the (always-open) input mode. Port of
// CaptureConnector.set_mode.
func (c *CaptureConnector) SetMode(mode string) []Frame {
	return []Frame{{"type": "worker_hello", "input_mode": "open"}}
}

// Clear resets buffer + pending and returns an empty term frame. Port of
// CaptureConnector.clear.
func (c *CaptureConnector) Clear() []Frame {
	c.mu.Lock()
	c.buffer = ""
	c.pending = ""
	c.mu.Unlock()
	return []Frame{{"type": "term", "data": ""}}
}

// GetAnalysis returns a human-readable analysis string. Port of
// CaptureConnector.get_analysis.
func (c *CaptureConnector) GetAnalysis() string {
	c.mu.Lock()
	defer c.mu.Unlock()
	s := fmt.Sprintf(
		"CaptureConnector socket=%q connected=%t buffer_len=%d stdin_keystrokes=%d outbound_connections=%d",
		c.socketPath, c.connected, len(c.buffer), c.stdinCount, len(c.connectLog),
	)
	if len(c.connectLog) > 0 {
		s += fmt.Sprintf(" recent_connect=%q", c.connectLog[len(c.connectLog)-1])
	}
	return s
}

// snapshotLocked builds a snapshot frame. Caller must hold c.mu. Port of
// CaptureConnector._snapshot.
func (c *CaptureConnector) snapshotLocked() Frame {
	screen := c.buffer
	return Frame{
		"type":               "snapshot",
		"screen":             screen,
		"cursor":             map[string]int{"row": 0, "col": 0},
		"cols":               c.cols,
		"rows":               c.rows,
		"screen_hash":        md5Hex(screen),
		"cursor_at_end":      true,
		"has_trailing_space": false,
		"prompt_detected":    false,
		"ts":                 nowTS(),
	}
}

File: packages/provide-uterm-platform/src/provide/uterm/pty/capture_connector.py
#
# SPDX-FileCopyrightText: Copyright (c) 2025-2026 provide.io llc. All rights reserved.
# SPDX-License-Identifier: AGPL-3.0-or-later
#
"""
CaptureConnector — session connector fed by libuterm_capture.so.

No process is forked.  Instead the connector listens on a Unix socket that
``libuterm_capture.so`` (injected via LD_PRELOAD) connects to at shell startup.
Incoming frames are accumulated into a text buffer and exposed as snapshots.

Only CHANNEL_STDOUT (0x01) contributes to the visible screen; CHANNEL_STDIN
(0x02) and CHANNEL_CONNECT (0x03) frames are recorded in the analysis log.

Config keys accepted in connector_config:
  socket_path       str    required — path of the Unix socket to listen on
                           (pam_uterm.so writes this as /run/uterm-cap-{pid}.sock)
  cols              int    terminal width hint (default 80)
  rows              int    terminal height hint (default 24)
  connect_timeout_s float  seconds to wait for capture lib to connect (default 5.0)
  stdin_socket_path str    optional — Unix socket path to forward browser keystrokes
                           to.  When set, handle_input() writes typed bytes there so
                           a listener can pipe them into the captured process's stdin.
"""

from __future__ import annotations

import asyncio
import hashlib
import time
from typing import Any

from provide.uterm.pty.capture import (
    CHANNEL_CONNECT,
    CHANNEL_STDIN,
    CHANNEL_STDOUT,
    CaptureSocket,
)

_VALID_CONFIG_KEYS = frozenset(
    {
        "socket_path",
        "cols",
        "rows",
        "connect_timeout_s",
        "input_mode",
        "stdin_socket_path",
    }
)


def _register() -> None:
    try:
        from provide.uterm.server.connectors.registry import register_connector

        register_connector("pty_capture", CaptureConnector)  # type: ignore[arg-type]  # ty:ignore[invalid-argument-type]
    except ImportError:
        pass


class CaptureConnector:
    """
    SessionConnector impl that observes an LD_PRELOAD-captured shell.

    connector_type = "pty_capture"
    """

    def __init__(self, session_id: str, display_name: str, config: dict[str, Any]) -> None:
        unknown = set(config) - _VALID_CONFIG_KEYS
        if unknown:
            raise ValueError(f"unknown config keys for CaptureConnector: {sorted(unknown)}")
        if "socket_path" not in config:
            raise ValueError("CaptureConnector requires 'socket_path' in connector_config")

        self._session_id = session_id
        self._display_name = display_name
        self._socket_path: str = str(config["socket_path"])
        self._cols: int = int(config.get("cols", 80))
        self._rows: int = int(config.get("rows", 24))
        self._connect_timeout: float = float(config.get("connect_timeout_s", 5.0))
        self._stdin_socket_path: str | None = (
            str(config["stdin_socket_path"]) if config.get("stdin_socket_path") else None
        )

        self._capture: CaptureSocket | None = None
        self._connected = False
        self._buffer = ""
        self._pending = ""  # new bytes not yet streamed to the browser
        self._connect_log: list[str] = []
        self._stdin_count = 0
        self._stdin_writer: asyncio.StreamWriter | None = None

    async def start(self) -> None:
        self._capture = CaptureSocket(self._socket_path)
        await self._capture.start()
        self._connected = True

    async def stop(self) -> None:
        await self._close_stdin_writer()
        if self._capture is not None:
            await self._capture.stop()
            self._capture = None
        self._connected = False

    def is_connected(self) -> bool:
        return self._connected

    async def poll_messages(self) -> list[dict[str, Any]]:
        if not self._connected or self._capture is None:
            return []
        changed = False
        # Drain all immediately available frames without blocking
        while True:
            frame = self._capture.read_nowait()
            if frame is None:
                break
            if frame.channel == CHANNEL_STDOUT:
                raw = frame.data.decode("utf-8", errors="replace")
                # Normalize bare \n → \r\n: DYLD capture bypasses the PTY ONLCR
                # driver, so xterm.js would advance cursor down without a CR.
                text = raw.replace("\r\n", "\n").replace("\n", "\r\n")
                self._buffer += text
                if len(self._buffer) > 65536:
                    self._buffer = self._buffer[-65536:]
                self._pending += text
                changed = True
            elif frame.channel == CHANNEL_STDIN:
                self._stdin_count += 1
            elif frame.channel == CHANNEL_CONNECT:
                addr = frame.data.decode("utf-8", errors="replace")
                self._connect_log.append(addr)
                if len(self._connect_log) > 100:
                    self._connect_log = self._connect_log[-100:]
        if changed and self._pending:
            data, self._pending = self._pending, ""
            return [{"type": "term", "data": data}]
        return []

    async def handle_input(self, data: str) -> list[dict[str, Any]]:
        if self._stdin_socket_path:
            await self._forward_stdin(data.encode("utf-8", errors="replace"))
        return []

    async def _forward_stdin(self, data: bytes) -> None:
        """Forward keystrokes to the stdin socket over an asyncio Unix stream.

        Lazy-connects, then writes + drains; on error reconnects and retries once.
        An asyncio stream (rather than a blocking ``socket.sendall``) keeps a slow
        or dead peer from stalling the event loop on a keystroke.
        """
        for _attempt in range(2):
            if self._stdin_writer is None:
                try:
                    _reader, self._stdin_writer = await asyncio.open_unix_connection(self._stdin_socket_path)
                except OSError:
                    return
            try:
                self._stdin_writer.write(data)
                await self._stdin_writer.drain()
                return
            except OSError:
                await self._close_stdin_writer()

    async def _close_stdin_writer(self) -> None:
        """Close the stdin stream writer, swallowing teardown errors."""
        if self._stdin_writer is None:
            return
        self._stdin_writer.close()
        try:
            await self._stdin_writer.wait_closed()
        except OSError:
            pass
        self._stdin_writer = None

    async def handle_control(self, action: str) -> list[dict[str, Any]]:
        return []

    async def get_snapshot(self) -> dict[str, Any]:
        return self._snapshot()

    async def set_mode(self, mode: str) -> list[dict[str, Any]]:
        return [{"type": "worker_hello", "input_mode": "open"}]

    async def clear(self) -> list[dict[str, Any]]:
        self._buffer = ""
        self._pending = ""
        return [{"type": "term", "data": ""}]

    async def get_analysis(self) -> str:
        return (
            f"CaptureConnector socket={self._socket_path!r} "
            f"connected={self._connected} buffer_len={len(self._buffer)} "
            f"stdin_keystrokes={self._stdin_count} "
            f"outbound_connections={len(self._connect_log)}"
            + (f" recent_connect={self._connect_log[-1]!r}" if self._connect_log else "")
        )

    def _snapshot(self) -> dict[str, Any]:
        screen = self._buffer
        return {
            "type": "snapshot",
            "screen": screen,
            # x/y, not row/col: that is what the snapshot frame contract and
            # every other connector use, and what the terminal element reads.
            "cursor": {"x": 0, "y": 0},
            "cols": self._cols,
            "rows": self._rows,
            # Non-cryptographic change-detection hash; `usedforsecurity=False`
            # is the canonical way to tell bandit/ruff and Python itself
            # that this is not a security boundary. Drops the orphan `nosec`
            # annotation that previously didn't match any active rule id.
            "screen_hash": hashlib.md5(screen.encode(), usedforsecurity=False).hexdigest(),
            "cursor_at_end": True,
            "has_trailing_space": False,
            # None, not False. The wire contract is a dict of detector output or
            # nothing at all, so a bool fails validation in the hub's frame
            # builder and the snapshot is dropped before any browser sees it —
            # which left a hijacked terminal blank until something repainted it.
            "prompt_detected": None,
            "ts": time.time(),
        }


_register()
